src/connectors/impls/s3_source.rs

```diff
@@ -0,0 +1,320 @@
+pub struct S3SourceConfig {
+    #[serde(default = "S3SourceConfig::default_access_key_id")]
+    aws_access_key_id: String,
+    #[serde(default = "S3SourceConfig::default_secret_token")]
+    aws_secret_access_key: String,
+    #[serde(default = "S3SourceConfig::default_aws_region")]
+    aws_region: String,
+
+    max_connections: usize,
+    endpoint: Option<String>,
+
+    bucket: String,
+    prefix: Option<String>,
+    delimiter: Option<String>,
+}
+
+// FIXME: insert better name here
+struct KeyPayload {
+    object_data: Object,
+    stream: u64,
+}
+
+// Defaults for the config.
+impl S3SourceConfig {
+    fn default_access_key_id() -> String {
+        "AWS_ACCESS_KEY_ID".to_string()
+    }
+
+    fn default_secret_token() -> String {
+        "AWS_SECRET_ACCESS_KEY".to_string()
+    }
+
+    fn default_aws_region() -> String {
+        "AWS_REGION".to_string()
+    }
+}
+
+impl ConfigImpl for S3SourceConfig {}
+
+#[derive(Debug, Default)]
+pub(crate) struct Builder {}
+
+#[async_trait::async_trait]
+impl ConnectorBuilder for Builder {
+    fn connector_type(&self) -> ConnectorType {
+        "s3-source".into()
+    }
+
+    async fn from_config(
+        &self,
+        id: &str,
+        raw_config: &Option<OpConfig>,
+    ) -> Result<Box<dyn Connector>> {
+        if let Some(config) = raw_config {
+            let mut config = S3SourceConfig::new(config)?;
+
+            // Fetch the secrets from the env.
+            config.aws_secret_access_key = env::var(config.aws_secret_access_key)?;
+            config.aws_access_key_id = env::var(config.aws_access_key_id)?;
+            config.aws_region = env::var(config.aws_region)?;
+
+            // Check the validity of given url
+            let endpoint = if let Some(url) = &config.endpoint {
+                let url_parsed = url.parse::<http::Uri>()?;
+                Some(url_parsed)
+            } else {
+                None
+            };
+
+            Ok(Box::new(S3SourceConnector {
+                handles: Vec::with_capacity(config.max_connections),
+                config,
+                endpoint,
+                tx: None,
+            }))
+        } else {
+            Err(ErrorKind::MissingConfiguration(id.to_string()).into())
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Connector for S3SourceConnector {
+    async fn create_source(
+        &mut self,
+        source_context: SourceContext,
+        builder: SourceManagerBuilder,
+    ) -> Result<Option<SourceAddr>> {
+        let (tx, rx) = channel::bounded(QSIZE.load(Ordering::Relaxed));
+        let s3_source = ChannelSource::from_channel(tx.clone(), rx);
+
+        self.tx = Some(tx);
+
+        let addr = builder.spawn(s3_source, source_context)?;
+        Ok(Some(addr))
+    }
+
+    async fn connect(&mut self, _ctx: &ConnectorContext, _attemp: &Attempt) -> Result<bool> {
+        let s3_config = s3::config::Config::builder()
+            .credentials_provider(Credentials::new(
```

Comment:
instead of this we should use: https://docs.rs/aws-config/latest/aws_config/fn.load_from_env.html which is instantiating this: https://docs.rs/aws-config/latest/aws_config/default_provider/credentials/struct.DefaultCredentialsChain.html

So the only thing that needs to go into the config for credentials, is the `region` being used.